Lib/test/test_os.py

```diff
@@ -2873,6 +2874,47 @@ def test_getfinalpathname_handles(self):
 
         self.assertEqual(0, handle_delta)
 
+    @support.requires_subprocess()
+    def test_stat_unlink_race(self):
+        # bpo-46785: the implementation of os.stat() falls back to reading
+        # the parent directory if CreateFileW() fails with a permission
+        # error. If reading the parent directory fails because the file or
+        # directory are subsequently unlinked, or because the volume or
+        # share are no longer available, then the original permission error
+        # should not be restored.
+        filename =  os_helper.TESTFN
+        self.addCleanup(os_helper.unlink, filename)
+        deadline = time.time() + 5
+        command = textwrap.dedent("""\
+            import os
+            import sys
+            import time
+            filename = sys.argv[1]
+            deadline = float(sys.argv[2])
+            while time.time() < deadline:
+                try:
+                    with open(filename, "w") as f:
+                        pass
+                except OSError:
+                    pass
+                try:
+                    os.remove(filename)
+                except OSError:
+                    pass
+            """)
+
+        with subprocess.Popen([sys.executable, '-c', command, filename, str(deadline)]):
+            while time.time() < deadline:
+                try:
+                    os.stat(filename)
+                # Note that `ERROR_NOT_READY`, which should also not be
+                # ignored, results in a `PermissionError`. That is not caught
+                # here, as only the behavior for `ERROR_FILE_NOT_FOUND` is
+                # checked in this test.
+                except FileNotFoundError:
+                    pass
```

Comment:
Terminate the child process if it got hung up for any reason:
 
```suggestion
        with subprocess.Popen([sys.executable, '-c', command, filename, str(deadline)]) as proc:
            while time.time() < deadline:
                try:
                    os.stat(filename)
                # Note that `ERROR_NOT_READY`, which should also not be
                # ignored, results in a `PermissionError`. That is not caught
                # here, as only the behavior for `ERROR_FILE_NOT_FOUND` is
                # checked in this test.
                except FileNotFoundError:
                    pass
            try:
                proc.wait(1)
            except subprocess.TimeoutExpired:
                proc.terminate()
```

Reply:
In fact, now I look at this, I think we have a helper for re-running ourselves which handles a range of environment issues we've run into in the past...

Reply:
Done.